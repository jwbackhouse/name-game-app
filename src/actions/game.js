// *** GAME ACTION CREATORS ***

import database from '../firebase/firebase';
import { getPlayersBegin, getPlayersSuccess, getPlayersFailure  } from './players';
import { addName, getNamesBegin, getNamesSuccess, getNamesFailure  } from './names';
import { addUserDetails } from './auth';


// Populate Firebase with initial game state
export const initialiseGame = () => {
  return () => {
    database.ref('game').set({
      teamAScore: 0,
      teamBScore: 0,
      playingNow: {
        uid: '',
        username: '',
        team: ''
      },
      startTime: '',
      startTurn: false,
      endGame: false,
      numPasses: 3,
      numNames: 5,
      timerLength: 60000,
    });
  };
};


// Subscribe to live Firebase data
export const fetchData = () => {
  return (dispatch, getState) => {
    dispatch(getPlayersBegin());
    dispatch(getGameDataBegin());
    dispatch(getNamesBegin());

    database.ref('players').on('value', snapshot => {
      let playersArr = [];
      snapshot.forEach(childSnapshot => {
        const playerData = childSnapshot.val();
        playersArr.push({
          uid: childSnapshot.key,
          ...playerData,
        });
        
        // Pass this player's existing data to the auth state object
        if (playerData.uid === getState().auth.uid) {
          dispatch(addUserDetails({
            playersUid: childSnapshot.key,
            ...playerData,
          }));
        };
      });
      dispatch(getPlayersSuccess(playersArr));
    }, err => dispatch(getPlayersFailure(err)));
      
    database.ref('game').on('value', snapshot => {
      const gameData = snapshot.val();
      dispatch(getGameDataSuccess(gameData));
    }, err => dispatch(getGameDataFailure(err)));
    
    database.ref('names').on('value', snapshot => {
      let namesArr = [];
      snapshot.forEach((childSnapshot) => {
        const nameData = childSnapshot.val();
        namesArr.push({
          id: childSnapshot.key,
          ...nameData
        });
      });
      dispatch(getNamesSuccess(namesArr));
    }, err => dispatch(getNamesFailure(err)));
  };
};

const getGameDataBegin = () => ({
  type: 'GET_GAME_DATA_BEGIN'
});

const getGameDataSuccess = gameData => ({
    type: 'GET_GAME_DATA_SUCCESS',
    payload: gameData
});

const getGameDataFailure = error => ({
  type: 'GET_GAME_DATA_FAILURE',
  payload: {error}
});

export const endFetchData = () => {
  return (dispatch, getState) => {
    database.ref('players').off();
    database.ref('game').off();
    database.ref('names').off();
  };
};


// Fetch scores
export const fetchScores = () => {
  return (dispatch) => {
    return database.ref('game').once('value')
      .then(snapshot => {
        const teamAScore = snapshot.val().teamAScore;
        const teamBScore = snapshot.val().teamBScore;
        dispatch(updateLocalScore(teamAScore, teamBScore))
      })
      .catch(err => console.log('actions/fetchScores(): error fetching game data from Firebase.'));
  };
};
      

// endGame flag
export const fetchEndGame = () => {
  return (dispatch, getState) => {
    database.ref('game/endGame').on('value', snapshot => {
      const endGame = snapshot.val();
      dispatch(updateEndGame(endGame));
    });
  };
};

export const endFetchEndGame = () => {
  return (dispatch, getState) => {
    database.ref('game/endGame').off();
  };
}

export const updateEndGame = (endGame) => ({
  type: 'UPDATE_END_GAME',
  endGame
});


// Update score
export const startUpdateScore = (team, score) => {
  return (dispatch, getState) => {
    const previousScore = team === 'A' ? getState().game.teamAScore : getState().game.teamBScore;
    // Check if previousScore exists
    const newScore = previousScore ? previousScore + score : score;
    const teamName = 'team' + team + 'Score';
    return database.ref('game').update({ [teamName]: newScore })
      .catch(err => console.log('Error updating score in Firebase:', err));
  };
};

export const updateLocalScore = (teamAScore, teamBScore) => ({
  type:'UPDATE_SCORE',
  teamAScore,
  teamBScore
});


// Next player
export const setNextPlayer = (player) => {
  return (dispatch) => {
    return database.ref(`game`).update({
      playingNow : {
        uid: player.uid,
        username: player.username,
        team: player.team
      }
    });
  };
};

export const resetNextPlayer = () => {
  return (dispatch) => {
    return database.ref(`game`).update({
      playingNow : {
        uid: '',
        username: '',
        team: ''
      }
    });
  };
};


// Players' turns
export const startTurn = () => {
  return () => {
    database.ref('game').update({
      startTurn: true,
      startTime: Date.now()
    });
  };
};

export const endTurn = (uid) => {
  return () => {
    database.ref().update({
      ['game/startTurn']: false,
      ['game/startTime']: '',
      [`players/${uid}/hasPlayed`]: true
    });
  };
};


// End game
export const endGame = (uid) => {
  return () => {
    database.ref().update({
      ['game/startTurn']: false,
      ['game/endGame']: true,
      ['game/startTime']: '',
      [`players/${uid}/hasPlayed`]: true
    });
  };
};

export const resetGame = () => ({
  type: 'RESET_GAME'
});