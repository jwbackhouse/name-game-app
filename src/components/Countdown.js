import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import database from '../firebase/firebase';

export class Countdown extends React.Component {
  state = {
    timerOn: false,
    timerTime: 0,
    timerStart: 0,
    timerLength: this.props.game.timerLength * 1000,
    message: '',
  }
  
  componentDidMount = () => {
    // Check for start time on Firebase
    database.ref('game/startTime').on('value', snapshot => {
      const startTime = snapshot.val();
      if (startTime) {
        this.setState({ timerStart: startTime, message: '' });
        this.startTimer();
      } else {
        this.setState({ message: 'Hang on a second...' });
      }
    });
  }
  
  componentWillUnmount = () => {
    database.ref('game/startTime').off();
    this.stopTimer();
  }
  
  startTimer = () => {
    this.setState({
      timerOn: true,
    });
    
    this.timer = setInterval(() => {
      const timeLeft = this.state.timerLength - this.state.timerTime;
      if (timeLeft >= 0)  {
        this.setState({ timerTime: Date.now() - this.state.timerStart });
      } else {
        clearInterval(this.timer);
        this.setState({ timerOn: false });
        this.props.onFinished();
      }
    }, 50);
  }

  stopTimer = () => {
    clearInterval(this.timer);
    this.setState({ timerOn: false });
  };
  
  // // Unused functionality
  // resetTimer = () => {
  //   this.setState({
  //     timerTime: 0,
  //     timerStart: 0
  //   });
  // };
  
  
  render = () => {
    const { timerLength, timerTime, message } = this.state;
    const timeLeft = Math.max(0, timerLength - timerTime);
    const messageElement = message && message;
    return (
      <div>
        <p>{ messageElement }</p>
        <h3>Time left: {moment(timeLeft).format('m:ss')}</h3>
      </div>
    )
  }
};

const mapStateToProps = (state) => ({
  game: state.game,
});

export default connect(mapStateToProps)(Countdown);
