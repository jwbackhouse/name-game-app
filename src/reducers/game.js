// *** GAME REDUCER ***

const initialState = {
  teamAScore: 0,
  teamBScore: 0,
  playingNow: {
    uid: '',
    userName: '',
    team: ''
  },
  startTime: undefined,
  startTurn: false,
  endGame: false
};

export default (state = initialState, action) => {
  switch(action.type) {
    case 'UPDATE_SCORE':
      return {
        ...state,
        teamAScore: action.teamAScore,
        teamBScore: action.teamBScore
      };
    case 'SET_PLAYING_TEAM':
      return {
        ...state,
        playingTeam: action.playingTeam
      };
    case 'UPDATE_GAME':
      return {
        ...state,
        ...action.gameData
      };
    case 'UPDATE_END_GAME':
      return {
        ...state,
        endGame: action.endGame
      };
    case 'RESET_GAME':
      return initialState;
    default:
      return state;
  };
};
