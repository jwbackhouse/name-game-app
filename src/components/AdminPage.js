import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { Link, useHistory } from 'react-router-dom';
import database from '../firebase/firebase';
import withLiveData from '../helpers/withLiveData';
import { resetGame, initialiseGame } from '../actions/game';
import { removeAllNames } from '../actions/names';


const AdminPage = ({ initialiseGame, removeAllNames, resetGame, game, history }) => {
  const [submitMsg, setSubmitMsg] = useState('');
  
  // Number of passes allowed per player
  const [numPasses, setNumPasses] = useState(game.numPasses || 2);
  const onPassesChange = (e) => setNumPasses(+e.target.value);
  useEffect(() => {
    if (game.numPasses !== undefined) setNumPasses(game.numPasses)  // needed otherwise input becomes uncontrolled
  }, [game.numPasses]);
  
  // Length of timer for guesses
  const [timerLength, setTimerLength] = useState(game.timerLength || 60000);
  const onTimerChange = (e) => {
    const milliseconds = e.target.value * 1000;
    setTimerLength(milliseconds);
  };
  
  // Very basic password protection for page access
  const [access, setAccess] = useState(undefined);
  if (access === undefined ) {
    const password = prompt('Please enter your password to view this page');
    const pass1 = 'namegame';
    password === pass1 ? setAccess(true): setAccess(false);
  }
  
  const onReset = () => {
    const confirm = window.confirm('Are you sure you want to reset the game? This can\'t be undone...');
    if (confirm) {
      database.ref().remove()
        .then(() => {
          initialiseGame();
          setSubmitMsg('Data successfully reset.');
        });
      removeAllNames();
      resetGame();
    }
  }
  
  const onSubmit = (e) => {
    e.preventDefault();
    
    database.ref('game').update({
      numPasses,
      timerLength,
    });
    setSubmitMsg('All saved for you.');
  };
  
  const pageContent = (
    <div>
      <h1>Game settings</h1>
      <form onSubmit={ onSubmit }>
        <label>Number of passes allowed
          <input
            name='numPasses'
            id='numPasses'
            value={ numPasses }
            onChange={ onPassesChange }
            type='number'
          />
        </label>
        <label>Time allowed for guessing (in seconds)
          <input
            name='timerLength'
            id='timerLength'
            value={ timerLength / 1000 }
            onChange={ onTimerChange }
            type='number'
          />
        </label>
        <button type='submit'>
          Save
        </button>
      </form>
      { submitMsg && <p>{ submitMsg }</p> }
      <br />
      <button onClick={ onReset }>
        Reset the game
      </button>
      <br />
      <br />
      <Link
        to='/'
        className='button'
      >
        Go home
      </Link>
    </div>
  );
  
  return access ? pageContent : null;
};

const mapStateToProps = (state) => ({
  game: state.game,
});

const mapDispatchToProps = (dispatch) => ({
  removeAllNames: () => dispatch(removeAllNames()),
  resetGame: () => dispatch(resetGame()),
  initialiseGame: () => dispatch(initialiseGame())
});


const connectedWithLiveData = compose(
  connect(mapStateToProps, mapDispatchToProps),
  withLiveData,
);

export default connectedWithLiveData(AdminPage);
