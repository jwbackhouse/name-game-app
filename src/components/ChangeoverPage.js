import React from 'react';
import { connect } from 'react-redux';
import StartMessage from './StartMessage';
import { fetchData, endFetchData, startTurn } from '../actions/game';
import { getNames } from '../actions/names';
import selectPlayer from '../selectors/selectPlayer';



export class ChangeoverPage extends React.Component {
  state = {
    error: '',
    nextPlayer: undefined
  }
  
  componentDidMount = () => {
    this.props.fetchData();
    this.props.getNames();
  }
  
  componentDidUpdate = (prevProps) => {
    if (!prevProps.game.startTurn) {
      this.props.game.startTurn && this.props.history.push('/guess');
    }
    if (this.props.game.endGame) {
      this.props.history.push('/end')
    }
  }
  
  componentWillUnmount = () => {
    this.props.endFetchData();
  }

  onClick = () => {
    // Update Firebase with startGame & startTime
    this.props.startTurn();
    this.props.history.push('/play')
  }
  
  render = () => {
    // Check if this user is playing next
    const thisUserPlaying = this.props.user.uid === this.props.game.playingNow.uid;

    return (
      <div className='content-container'>
        <h1>The scores on the doors</h1>
        <p>Team A: {this.props.game.teamAScore}</p>
        <p>Team B: {this.props.game.teamBScore}</p>
        <StartMessage
          thisUserPlaying={ thisUserPlaying }
          playingNow={ this.props.game.playingNow }
          errorMsg={ '' }
          onClick={ this.onClick }
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  players: state.players,
  game: state.game,
  user: state.user
});

const mapDispatchToProps = (dispatch) => ({
  startTurn: () => dispatch(startTurn()),
  fetchData: () => dispatch(fetchData()),
  getNames: () => dispatch(getNames()),
  endFetchData: () => dispatch(endFetchData())
});


export default connect(mapStateToProps, mapDispatchToProps)(ChangeoverPage);

