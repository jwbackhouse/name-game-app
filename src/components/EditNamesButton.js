import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { Link } from 'react-router-dom';
import { togglePlayerReady } from '../actions/players';

export const EditNamesButtonBase = props => {
  const onTogglePlayerReady = () => {
    props.togglePlayerReady(props.auth.uid, props.auth.firebaseUID);
    props.history.push('/setup');
  };
  
  return (
    <button className='button button--background' onClick={ onTogglePlayerReady }>
      Edit your names
    </button>
  );
};

const mapStateToProps = state => ({
  auth: state.auth
});

const mapDispatchToProps = dispatch => ({
  togglePlayerReady: (uid, playersUid) => dispatch(togglePlayerReady(uid, playersUid)),
})

// withRouter required to access history
const EditNamesButton = withRouter(EditNamesButtonBase);
export default connect(mapStateToProps, mapDispatchToProps)(EditNamesButton);
