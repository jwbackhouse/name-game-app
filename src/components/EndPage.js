import React from 'react';
import { connect } from 'react-redux';
import database from '../firebase/firebase';
import { fetchData } from '../actions/game';
import { removeAllNames } from '../actions/names';


export class EndPage extends React.Component {
  componentDidMount = () => {
    this.props.fetchData();
  }
    
  onClick = () => {
    // Clear database
    database.ref().remove();
    // Necessary as db update on Register page doesn't over-write names
    this.props.removeAllNames();
    this.props.history.push('/');
  }
  
  render = () => {
    let message;
    if (this.props.game.teamAScore > this.props.game.teamBScore) {
      message = 'Team A has it!';
    } else if (this.props.game.teamBScore > this.props.game.teamAScore) {
      message = 'Team B has it!';
    } else {
      message = "It's a draw!";
    }
    
    return (
      <div className='content-container'>
        <h1>{ message }</h1>
        <p>Team A scored { this.props.game.teamAScore }</p>
        <p>Team B scored { this.props.game.teamBScore }</p>
        <button onClick={ this.onClick }>Play again</button>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  game: state.game,
});

const mapDispatchToProps = (dispatch) => ({
  fetchData: () => dispatch(fetchData()),
  removeAllNames: () => dispatch(removeAllNames())
});

export default connect(mapStateToProps, mapDispatchToProps)(EndPage);

