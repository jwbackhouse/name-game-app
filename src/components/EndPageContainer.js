import React from 'react';
import { connect } from 'react-redux';
import database from '../firebase/firebase';
import EndPage from './EndPage';
import { resetGame, initialiseGame, fetchScores } from '../actions/game';
import { removeAllNames } from '../actions/names';


export class EndPageContainer extends React.Component {
  state = {
    ready: false
  }
  
  componentDidMount() {
    this.props.fetchScores()
      .then(() => this.setState({ ready: true }));
  }
    
  onClick = () => {
    const { initialiseGame, removeAllNames, resetGame, history } = this.props;
    
    // Clear database & initialise new game
    database.ref().remove()
      .then(() => initialiseGame());
      
    // Clear redux store
    removeAllNames();
    resetGame();
    
    history.push('/');
  }
  
  render = () => {
    const { teamAScore, teamBScore } = this.props.game;
    
    let message;
    if (teamAScore > teamBScore) {
      message = 'Team A has it!';
    } else if (teamAScore < teamBScore) {
      message = 'Team B has it!';
    } else {
      message = "It's a draw!";
    }
    
    const renderComponent = this.state.ready && (
      <EndPage
          message={ message }
          teamAScore={ teamAScore }
          teamBScore={ teamBScore }
          onClick={ this.onClick }
      />
    )
    
    return <div>{ renderComponent }</div>
  }
}

const mapStateToProps = (state) => ({
  game: state.game,
});

const mapDispatchToProps = (dispatch) => ({
  fetchScores: () => dispatch(fetchScores()),
  removeAllNames: () => dispatch(removeAllNames()),
  resetGame: () => dispatch(resetGame()),
  initialiseGame: () => dispatch(initialiseGame())
});

export default connect(mapStateToProps, mapDispatchToProps)(EndPageContainer);