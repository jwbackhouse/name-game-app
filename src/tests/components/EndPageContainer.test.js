import React from 'react';
import { shallow, mount } from 'enzyme';
import database from '../../firebase/firebase';
import game from '../fixtures/testGame';
import { EndPageContainer} from '../../components/EndPageContainer';

// jest.mock('../../actions/game', () => {
//   return {
//     fetchScores: jest.fn(() => Promise.resolve()),
//   };
// });


describe('<EndPageContainer mounted />', () => {
  let wrapper;
  let EndPage;
  const fetchScores = jest.fn(() => Promise.resolve(''));
  const initialiseGame = jest.fn();
  const removeAllNames = jest.fn();
  const resetGame = jest.fn();
  const history = { push: jest.fn() };

  beforeEach(() => {
    wrapper = mount(
      <EndPageContainer
        fetchScores={ fetchScores }
        initialiseGame={ initialiseGame }
        removeAllNames={ removeAllNames }
        resetGame={ resetGame }
        game={ game }
        history = { history}
      />
    );
  });
  
  afterEach(() => {
    jest.clearAllMocks();
  });
  
  
  test('Should render empty component', () => {
    // State is set as ready, but think something about how Jest/Enzyme works is running snapshot before EndPage renders
    expect(wrapper).toMatchSnapshot();
    expect(fetchScores).toBeCalled();
    expect(wrapper.state().ready).toBe(true);
  });
  
  test('Should render EndPage if state set to ready: true', () => {
    wrapper.setState({ ready: true });
    const EndPage = wrapper.find('EndPage');
    expect(EndPage.exists()).toBe(true);
  });
  
  test('Should pass correct values to props for EndGame component', () => {
    wrapper.setState({ ready: true });
    const EndPage = wrapper.find('EndPage');
    expect(EndPage.props().teamAScore).toEqual(game.teamAScore);
    expect(EndPage.props().teamBScore).toEqual(game.teamBScore);
  });
  
  test('Should handle button click correctly.', () => {
    // // Need to mock database.ref().remove() in order to trigger initialiseGame
    // const promise = new Promise((resolve, reject) => {
    //   resolve('abc');
    //   done();
    // });
    // database.ref.remove = jest.fn(() => promise);
    // expect(initialiseGame).toBeCalled();
    
    wrapper.setState({ ready: true });
    wrapper.find('button').simulate('click');
    // Add delay to allow Firebase query to complete
    setTimeout(() => {
      expect(removeAllNames).toBeCalled();
      expect(resetGame).toBeCalled();
      expect(history.push).toBeCalledWith('/');
    },500)
  });
});
