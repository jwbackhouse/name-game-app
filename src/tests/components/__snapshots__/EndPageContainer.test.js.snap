// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EndPageContainer mounted /> Should render empty component 1`] = `
<EndPageContainer
  fetchScores={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  }
  game={
    Object {
      "endGame": false,
      "error": "",
      "isLoading": false,
      "numPasses": 1,
      "playingNow": Object {
        "team": "B",
        "uid": 1234,
        "userName": "Bob",
      },
      "startTime": undefined,
      "startTurn": false,
      "teamAScore": 4,
      "teamBScore": 2,
      "timerLength": 30000,
    }
  }
  history={
    Object {
      "push": [MockFunction],
    }
  }
  initialiseGame={[MockFunction]}
  removeAllNames={[MockFunction]}
  resetGame={[MockFunction]}
>
  <div />
</EndPageContainer>
`;
